### PR TITLE
08108 phased fractional timer

### LINKED ISSUE #8108: Phased Fractional Timer
It would be nice if we had a utility for for measuring the fraction of time that a thread (or whatever) spends in different phases of operation. For example, when attempting to measure the fraction of time that a thread is spending in different phases of execution.

### PR DESCRIPTION
**Pull request by cody-littley**
Closes #8108

### CODE REVIEW

**Review by edward-swirldslabs — DISMISSED** [review]

**Review by hendrikebbers — APPROVED** [review]

File: platform-sdk/swirlds-common/src/main/java/com/swirlds/common/metrics/extensions/FractionalTimer.java [hunk 1]
```diff
@@ -0,0 +1,288 @@
...
+     *                         inactive
+     * @param now              the current time in microseconds
+     * @return the new value that will be stored in the atomic long.
+     */
+    private long statusUpdate(final long currentState, final boolean isBecomingActive, final long now) {
+
+        // the epoch time when the last reset occurred
+        final int measurementStart = ByteUtils.extractLeftInt(currentState);
+        // The current status is represented by the (+ -) sign. The number represents the time spent in the
+        // opposite status. This is that its time spent being active/inactive can be deduced whenever the sample is
```
Comment (alittley) [inline platform-sdk/swirlds-common/src/main/java/com/swirlds/common/metrics/extensions/FractionalTimer.java:225]: this is *so* that its time...

File: platform-sdk/swirlds-common/src/main/java/com/swirlds/common/metrics/extensions/PhaseTimer.java [hunk 2]
```diff
@@ -0,0 +1,178 @@
...
+ * @param <T> the type of the phase, must be an enum
+ */
+public class PhaseTimer<T extends Enum<T>> {
+
+    private final Time time;
+
+    private final boolean fractionMetricsEnabled;
+    private final boolean absoluteTimeMetricsEnabled;
+
+    private final Map<T, FractionalTimer> fractionalTimers = new HashMap<>();
```
Comment (alittley) [inline platform-sdk/swirlds-common/src/main/java/com/swirlds/common/metrics/extensions/PhaseTimer.java:43]: A short explanation of fractional vs absolute timers would be helpful

File: platform-sdk/swirlds-common/src/main/java/com/swirlds/common/metrics/extensions/PhaseTimerBuilder.java [hunk 3]
```diff
@@ -0,0 +1,230 @@
...
+
+    /**
+     * Set the initial phase. If not set, the phase with ordinal 0 is used.
+     *
+     * @param initialPhase the initial phase
+     * @return this
+     */
+    @NonNull
+    public PhaseTimerBuilder<T> setInitialPhase(@NonNull final T initialPhase) {
+        this.initialPhase = initialPhase;
```
Comment (alittley) [inline platform-sdk/swirlds-common/src/main/java/com/swirlds/common/metrics/extensions/PhaseTimerBuilder.java:101]: requireNonNull

File: platform-sdk/swirlds-common/src/main/java/com/swirlds/common/metrics/extensions/PhaseTimerBuilder.java [hunk 4]
```diff
@@ -0,0 +1,230 @@
...
+    }
+
+    /**
+     * Set the unit for absolute time metrics. If not set, microseconds are used.
+     *
+     * @param absoluteTimeUnit the unit for absolute time metrics
+     * @return this
+     */
+    @NonNull
+    public PhaseTimerBuilder<T> setAbsoluteTimeUnit(@NonNull final TimeUnit absoluteTimeUnit) {
```
Comment (alittley) [inline platform-sdk/swirlds-common/src/main/java/com/swirlds/common/metrics/extensions/PhaseTimerBuilder.java:134]: missed opportunity to call this `absoluteUnit`

File: platform-sdk/swirlds-unit-tests/common/swirlds-common-test/src/test/java/com/swirlds/common/test/metrics/extensions/FractionalTimerTests.java [hunk 5]
```diff
@@ -41,71 +43,71 @@ void reset() {
...
                 0.43,
-                metric.getBusyFraction(),
+                metric.getActiveFraction(),
                 0.01,
                 "3 second of work, 4 seconds of idle, so the busy fraction should be 0.43");
         assertEquals(
                 0.43,
                 metric.getAndReset(),
                 0.01,
                 "the snapshot should contain the same value returned by getBusyFraction()");
```
Comment (alittley) [inline platform-sdk/swirlds-unit-tests/common/swirlds-common-test/src/test/java/com/swirlds/common/test/metrics/extensions/FractionalTimerTests.java:104]: assertion messages throughout this class mention old method names

File: platform-sdk/swirlds-unit-tests/common/swirlds-common-test/src/test/java/com/swirlds/common/test/metrics/extensions/PhaseTimerTests.java [hunk 6]
```diff
@@ -0,0 +1,592 @@
...
+
+        for (int i = 0; i < 100; i++) {
+            time.tick(Duration.ofSeconds(1));
+            timer.activatePhase(BAR);
+            time.tick(Duration.ofSeconds(1));
+            timer.activatePhase(BAZ);
+        }
+
+        assertEquals(2, fooUpdates.size());
+        assertEquals(fooUpdates.get(0), 3.0);
```
Comment (alittley) [inline platform-sdk/swirlds-unit-tests/common/swirlds-common-test/src/test/java/com/swirlds/common/test/metrics/extensions/PhaseTimerTests.java:274]: nit: you have `actual, expected` in many places, where it ought to be `expected, actual`

**Review by alittley — APPROVED** [review]

File: platform-sdk/swirlds-common/src/main/java/com/swirlds/common/metrics/extensions/FractionalTimer.java (hunk 1 above)
Comment (cody-littley) [reply platform-sdk/swirlds-common/src/main/java/com/swirlds/common/metrics/extensions/FractionalTimer.java:225]: fixed

File: platform-sdk/swirlds-common/src/main/java/com/swirlds/common/metrics/extensions/PhaseTimer.java (hunk 2 above)
Comment (cody-littley) [reply platform-sdk/swirlds-common/src/main/java/com/swirlds/common/metrics/extensions/PhaseTimer.java:43]:
```
    /**
     * Tracks the fraction of time, out of 1.0, spent in each phase.
     */
    private final Map<T, FractionalTimer> fractionalTimers = new HashMap<>();

    /**
     * Tracks the average time spent in a phase before switching to another phase.
     */
    private final Map<T, RunningAverageMetric> absoluteTimeMetrics = new HashMap<>();
```

File: platform-sdk/swirlds-common/src/main/java/com/swirlds/common/metrics/extensions/PhaseTimerBuilder.java (hunk 3 above)
Comment (cody-littley) [reply platform-sdk/swirlds-common/src/main/java/com/swirlds/common/metrics/extensions/PhaseTimerBuilder.java:101]: fixed

File: platform-sdk/swirlds-common/src/main/java/com/swirlds/common/metrics/extensions/PhaseTimerBuilder.java (hunk 4 above)
Comment (cody-littley) [reply platform-sdk/swirlds-common/src/main/java/com/swirlds/common/metrics/extensions/PhaseTimerBuilder.java:134]: "fixed"

File: platform-sdk/swirlds-unit-tests/common/swirlds-common-test/src/test/java/com/swirlds/common/test/metrics/extensions/FractionalTimerTests.java (hunk 5 above)
Comment (cody-littley) [reply platform-sdk/swirlds-unit-tests/common/swirlds-common-test/src/test/java/com/swirlds/common/test/metrics/extensions/FractionalTimerTests.java:104]: fixed

File: platform-sdk/swirlds-unit-tests/common/swirlds-common-test/src/test/java/com/swirlds/common/test/metrics/extensions/PhaseTimerTests.java (hunk 6 above)
Comment (cody-littley) [reply platform-sdk/swirlds-unit-tests/common/swirlds-common-test/src/test/java/com/swirlds/common/test/metrics/extensions/PhaseTimerTests.java:274]: fixed

**Review by alittley — APPROVED** [review]

**Review by edward-swirldslabs — APPROVED** [review]

**Pull request merged**